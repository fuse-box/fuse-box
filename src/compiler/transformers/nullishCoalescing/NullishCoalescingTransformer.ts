import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { createUndefinedVariable, createVariableDeclaration } from '../../helpers/astHelpers';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { GlobalContext } from '../../program/GlobalContext';

export function createNullishStatement(props: ILocalContext) {
  const { genId, globalContext } = props;
  const expr: ASTNode = {
    expression: {
      alternate: undefined,
      consequent: undefined,
      test: {
        left: {
          left: undefined,
          operator: '!==',
          right: {
            type: 'Literal',
            value: null,
          },
          type: 'BinaryExpression',
        },
        operator: '&&',
        right: {
          left: undefined,
          operator: '!==',
          right: {
            argument: {
              type: 'Literal',
              value: 0,
            },
            operator: 'void',
            prefix: true,
            type: 'UnaryExpression',
          },
          type: 'BinaryExpression',
        },
        type: 'LogicalExpression',
      },
      type: 'ConditionalExpression',
    },
    type: 'ExpressionStatement',
  };
  return {
    expression: expr,
    setCondition: (node: ASTNode) => {
      // simple identifiers
      if (node.type === ASTType.Identifier) {
        expr.expression.consequent = node;
        expr.expression.test.left.left = node;
        expr.expression.test.right.left = node;
      } else {
        const sysVariable = genId();
        const target = { name: sysVariable, type: ASTType.Identifier };
        expr.expression.test.left.left = {
          left: target,
          operator: '=',
          right: node,
          type: 'AssignmentExpression',
        };
        expr.expression.test.right.left = target;
        expr.expression.consequent = target;
      }
    },
    setRight: (node: ASTNode) => {
      expr.expression.alternate = node;
    },
  };
}

function isNullishCoalescing(node: ASTNode) {
  return node.type === ASTType.LogicalExpression && node.operator === '??';
}

function drillExpressions(node: ASTNode, nodes: Array<ASTNode>) {
  if (!isNullishCoalescing(node)) return;
  nodes.unshift(node);
  if (node.left) drillExpressions(node.left, nodes);
}

interface ILocalContext {
  globalContext: GlobalContext;
  genId?: () => string;
}

export function NullishCoalescingTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const { node } = visit;

          const nodes: Array<ASTNode> = [];
          if (!isNullishCoalescing(node)) return;
          const globalContext = visit.globalContext as GlobalContext;
          drillExpressions(node, nodes);

          const startingNode = nodes[0];
          const declaration = createVariableDeclaration();

          const ctx: ILocalContext = {
            globalContext,
            genId: () => {
              const nextVar = globalContext.getNextSystemVariable();
              declaration.declarations.push(createUndefinedVariable(nextVar));
              return nextVar;
            },
          };

          let pointer = createNullishStatement(ctx);
          pointer.setRight(startingNode.right);
          pointer.setCondition(startingNode.left);
          let index = 1;
          while (index < nodes.length) {
            const item = nodes[index];
            const newPointer = createNullishStatement(ctx);
            newPointer.setRight(item.right);
            newPointer.setCondition(pointer.expression.expression);
            pointer = newPointer;
            index++;
          }
          let prepend: Array<ASTNode> = [];
          if (declaration.declarations.length) prepend = [declaration];

          return { prependToBody: prepend, replaceWith: pointer.expression.expression };
        },
      };
    },
  };
}
