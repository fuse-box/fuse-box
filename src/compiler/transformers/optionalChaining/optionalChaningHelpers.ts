import { ASTNode, ASTType } from '../../interfaces/AST';

export function ensureExpressionConsistency(expression: ASTNode, node?: ASTNode) {
  if (node.type == ASTType.Literal) {
    expression.computed = true;
  }
}

export function convertToCallExpression(alternate: ASTNode, id, args): ASTNode {
  return {
    arguments: args,
    callee: { object: { name: id, type: 'Identifier' }, property: alternate, type: 'MemberExpression' },
    type: 'CallExpression',
  };
}

export type OptionalChainHelper = ReturnType<typeof createOptionalChaningExpression>;
export function createOptionalChaningExpression(userId: string) {
  // prevMember.test.left.left.right

  // prevMement.alternate.property
  const AST_NODE: ASTNode = {
    alternate: null,
    consequent: {
      argument: {
        type: 'Literal',
        value: 0,
      },
      operator: 'void',
      prefix: true,
      type: 'UnaryExpression',
    },
    test: {
      left: {
        left: {
          left: {
            name: userId,
            type: 'Identifier',
          },
          operator: '=',
          right: null,
          type: 'AssignmentExpression',
        },
        operator: '===',
        right: {
          type: 'Literal',
          value: null,
        },
        type: 'BinaryExpression',
      },
      operator: '||',
      right: {
        left: {
          name: userId,
          type: 'Identifier',
        },
        operator: '===',
        right: {
          argument: {
            type: 'Literal',
            value: 0,
          },
          operator: 'void',
          prefix: true,
          type: 'UnaryExpression',
        },
        type: 'BinaryExpression',
      },
      type: 'LogicalExpression',
    },
    type: 'ConditionalExpression',
  };

  return {
    id: userId,
    statement: AST_NODE,
    setLeft: (item: ASTNode) => {
      AST_NODE.test.left.left.right = item;
    },
    setRight: (item: ASTNode) => {
      AST_NODE.alternate = item;
    },
  };
}
