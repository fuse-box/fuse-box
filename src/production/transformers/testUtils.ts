import { ITransformer } from '../../compiler/interfaces/ITransformer';
import * as buntis from 'buntis';
import { GlobalContextTransformer } from '../../compiler/transformers/GlobalContextTransformer';
import { transpileModule } from '../../compiler/program/transpileModule';
import { createGlobalContext } from '../../compiler/program/GlobalContext';
import { createContext } from '../../core/Context';
import { createPackage } from '../../core/Package';
import { createModule } from '../../core/Module';
import { ASTNode } from '../../compiler/interfaces/AST';
import { ProductionContext } from '../ProductionContext';
import { ModuleTree } from '../module/ModuleTree';

export function testProductionWarmup(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
  moduleProps?: any;
}) {
  const ctx = createContext({
    cache: false,
    devServer: false,
    target: 'browser',
  });

  const pkg = createPackage({ ctx: ctx, meta: {} as any });

  const module = Object.assign(
    createModule({
      absPath: __filename,
      ctx: ctx,
      extension: '.ts',
      fuseBoxPath: 'file.ts',
    }, pkg),
    props.moduleProps || {},
  );

  pkg.modules.push(module);

  module.contents = props.code;
  module.parse();

  const productionContext = ProductionContext(ctx, [pkg]);

  const tree = (module.moduleTree = ModuleTree(productionContext, module));

  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.productionWarmupPhase) {
      tranformers.push(t.productionWarmupPhase({ ctx: ctx, module: module, productionContext: productionContext }));
    }
  }
  transpileModule({
    ast: module.ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });

  return { productionContext, module, tree };
}
