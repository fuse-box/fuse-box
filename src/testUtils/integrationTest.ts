import * as fs from 'fs';
import { writeFileSync } from 'fs';
import * as fsExtra from 'fs-extra';
import * as path from 'path';
import { EnvironmentType } from '../config/EnvironmentType';
import { IPublicConfig } from '../config/IConfig';
import { Context, createContext } from '../core/context';
import { bundleDev, IBundleDevResponse } from '../development/bundleDev';
import { env } from '../env';
import { ensureDir, fastHash, fileExists, readFile } from '../utils/utils';
export interface ITestBrowserResponse {
  __fuse: any;
  entry: any;
  errors: Array<any>;
}
export function createDevSandbox(props: { ctx: Context; response: IBundleDevResponse; workspace: ITestWorkspace }) {
  const self = {
    runBundleInBrowser: async (): Promise<ITestBrowserResponse> => {
      const app = readFile(path.join(props.workspace.distRoot, 'app.js'));
      const window: any = {};
      const __fuse: any = new Proxy(
        {},
        {
          get: (obj, prop) => {
            return window.__fuse[prop];
          },
        },
      );
      var fn = new Function('window', '__fuse', app);
      const errors = [];
      try {
        fn(window, __fuse);
      } catch (e) {
        errors.push(e);
      }
      return new Promise((resolve, reject) => {
        return resolve({
          __fuse,
          entry: __fuse.r(1),
          errors,
        });
      });
    },
  };
  return self;
}

export type ITestWorkspace = ReturnType<typeof createTestWorkspace>;

function getCacheWorkspace(workspace: ITestWorkspace) {
  const files = fs.readdirSync(workspace.cacheFolder);

  const cacheFolder = path.join(workspace.cacheFolder, files[0]);
  const metaFile = path.join(cacheFolder, 'meta.json');
  const cacheFilesFolder = path.join(cacheFolder, 'files');

  return {
    getCachedFiles: () => {
      const files = fs.readdirSync(cacheFilesFolder);
      return files.map(i => path.join(cacheFilesFolder, i));
    },
    getMetaFile: () => {
      return JSON.parse(readFile(metaFile));
    },
  };
}
export function createTestWorkspace(props: {
  cache?: boolean;
  files: Record<string, string>;
  modules?: Record<string, Record<string, string>>;
}) {
  const sourceFolder = fastHash(Math.random() + '-' + new Date().getTime());
  const rootDir = path.join(env.APP_ROOT, '.tmp', sourceFolder);
  const sourceDir = path.join(rootDir, 'src');
  const distRoot = path.join(rootDir, 'dist');
  let cacheFolder;
  if (props.cache) {
    cacheFolder = path.join(rootDir, '.cache');
  }

  // creating modules if needed
  const modulesRoot = path.join(rootDir, 'modules');
  let modules = [];
  if (props.modules) {
    for (const name in props.modules) {
      for (const filename in props.modules[name]) {
        const filePath = path.join(modulesRoot, name, filename);
        ensureDir(path.dirname(filePath));
        writeFileSync(filePath, props.modules[name][filename]);
      }
    }
    modules.push(modulesRoot);
  }

  const webIndex = path.join(distRoot, 'index.html');

  // creating user files
  ensureDir(sourceDir);
  for (const filename in props.files) {
    const filePath = path.join(sourceDir, filename);
    ensureDir(path.dirname(filePath));
    writeFileSync(filePath, props.files[filename]);
  }
  const self = {
    cacheFolder,
    distRoot,
    modules,
    rootDir,
    sourceDir,
    webIndex,
    destroy: () => {
      fsExtra.removeSync(rootDir);
    },
    getCacheWorkspace: () => {
      return getCacheWorkspace(self);
    },
    removeFile: filename => {
      const filePath = path.join(sourceDir, filename);
      if (fileExists(filePath)) {
        fsExtra.removeSync(filePath);
      }
    },
    removeModuleFile: filename => {
      const filePath = path.join(rootDir, 'modules', filename);
      if (fileExists(filePath)) {
        fsExtra.removeSync(filePath);
      }
    },
    setFile: (filename: string, contents: string) => {
      const filePath = path.join(sourceDir, filename);
      ensureDir(path.dirname(filePath));
      writeFileSync(filePath, contents);
    },
    setModuleFile: (filename: string, contents: string) => {
      const filePath = path.join(rootDir, 'modules', filename);
      ensureDir(path.dirname(filePath));
      writeFileSync(filePath, contents);
    },
  };
  return self;
}
export function createIntegrationTest(props: {
  config?: IPublicConfig;
  envType: EnvironmentType;
  workspace: ITestWorkspace;
}) {
  const customConfig: IPublicConfig = props.config || {};
  const { workspace } = props;

  const config: IPublicConfig = {
    cache: { enabled: !!workspace.cacheFolder, root: workspace.cacheFolder },
    devServer: false,
    logging: { level: 'disabled' },
    modules: workspace.modules,
    watcher: false,
    webIndex: { publicPath: './' },
    ...customConfig,
  };

  const ctx = createContext({
    envType: props.envType,
    publicConfig: config,
    runProps: { bundles: { app: 'app.js', distRoot: workspace.distRoot } },
    scriptRoot: workspace.rootDir,
  });

  return {
    ctx,
    runDev: async () => {
      const response = await bundleDev({ ctx });

      return createDevSandbox({ ctx, response, workspace: props.workspace });
    },
  };
}
