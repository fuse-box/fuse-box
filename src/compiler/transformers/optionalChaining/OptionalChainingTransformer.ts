import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { createUndefinedVariable, createVariableDeclaration } from '../../helpers/astHelpers';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { OptionalChainHelper, createOptionalChaningExpression } from './optionalChaningHelpers';

const VALID_NODES = {
  [ASTType.OptionalCallExpression]: 1,
  [ASTType.OptionalMemberExpression]: 1,
};

const NON_DRILLABLE_NODES = {
  [ASTType.CallExpression]: 1,
  [ASTType.Identifier]: 1,
  [ASTType.MemberExpression]: 1,
};

interface IChainingStep {
  callArguments?: Array<ASTNode>;
  computed?: boolean;
  expression?: ASTNode;
  optional?: boolean;
}

interface IChainItem {
  items: Array<IChainingStep>;
}

type IChainingCollection = Array<IChainItem>;

function createOptionalContext(visit: IVisit) {
  const steps: Array<IChainingStep> = [];
  const globalContext = visit.globalContext;
  let finalStatement: ASTNode;
  let flatCollection: IChainingCollection;
  const declaration = createVariableDeclaration();

  const self = {
    declaration,
    finalStatement,
    flatCollection,
    steps,
    genId: () => {
      const nextVar = globalContext.getNextSystemVariable();
      declaration.declarations.push(createUndefinedVariable(nextVar));
      return nextVar;
    },
  };
  return self;
}
type OptionalChainContext = ReturnType<typeof createOptionalContext>;

function createPartialExpression(props: {
  flatExpression: IFlatExpression;
  nodes: Array<IChainingStep>;
  shouldExtractThis: boolean;
}) {
  let expression: ASTNode;
  const { flatExpression, nodes, shouldExtractThis } = props;
  const firstNode = nodes[0];
  let thisVariable: string;

  if (nodes.length === 1) {
    if (shouldExtractThis) {
      if (firstNode.expression.type === 'MemberExpression') {
        thisVariable = flatExpression.context.genId();
        firstNode.expression.object = {
          left: { name: thisVariable, type: 'Identifier' },
          operator: '=',
          right: firstNode.expression.object,
          type: 'AssignmentExpression',
        };
        return { expression: firstNode.expression, thisVariable };
      }
    }
    return { expression: firstNode.expression };
  }

  const total = nodes.length;
  let index = 0;

  while (index < total) {
    const item = nodes[index];
    const isLast = index === total - 1;

    const computed = item.computed || (item.expression && item.expression.type === ASTType.Literal);

    // call expression
    if (item.callArguments) {
      if (index === 1) {
        expression = createOptionalCall(flatExpression, item.callArguments);
      } else {
        expression = {
          arguments: item.callArguments,
          callee: expression,
          type: 'CallExpression',
        };
      }
    } else {
      if (!expression) {
        expression = {
          object: item.expression,
          property: null,
          type: 'MemberExpression',
        };
      } else {
        // if we see a next property while having an existing CallExpression, we must
        // converte to a MemberExpression with an object as the current CallExpression
        if (expression.type === 'CallExpression')
          expression = { object: expression, property: null, type: 'MemberExpression' };

        if (!expression.property) {
          expression.property = item.expression;
          expression.computed = computed;
        } else {
          expression = {
            computed: computed,
            object: expression,
            property: item.expression,
            type: 'MemberExpression',
          };
        }
      }
    }
    if (isLast && shouldExtractThis) {
      // adding an expression statement
      // E.g a statement like a?.b.c.d.e.f?.();
      // The second part b.c.d.e.f, gets converted into
      // (_new_variable = _1_.b.c.d.e).f)
      if (expression.object) {
        thisVariable = flatExpression.context.genId();
        expression.object = {
          left: { name: thisVariable, type: 'Identifier' },
          operator: '=',
          right: expression.object,
          type: 'AssignmentExpression',
        };
      } else {
      }
    }
    index++;
  }

  return { expression, thisVariable };
}

interface IFlatExpression {
  alternate?: Array<IChainingStep>;
  conditionSteps?: Array<IChainingStep>;
  context?: OptionalChainContext;
  id?: string;
  nextFlatExpression?: IFlatExpression;
  shouldExtractThis?: boolean;
  thisVariable?: string;
  collect?: () => OptionalChainHelper;
  generate?: () => OptionalChainHelper;
}

function createOptionalCall(expression: IFlatExpression, callArguments: Array<ASTNode>): ASTNode {
  let args: Array<ASTNode> = [];
  if (expression.thisVariable) {
    args.push({
      name: expression.thisVariable,
      type: 'Identifier',
    });
  } else {
    return {
      arguments: callArguments,
      callee: {
        computed: false,
        name: expression.id,
        type: 'Identifier',
      },
      type: 'CallExpression',
    };
  }
  args = args.concat(callArguments);
  return {
    arguments: args,
    callee: {
      computed: false,
      object: {
        name: expression.id,
        type: 'Identifier',
      },
      property: {
        name: 'call',
        type: 'Identifier',
      },
      type: 'MemberExpression',
    },
    type: 'CallExpression',
  };
}

function createFlatExpression(context: OptionalChainContext): IFlatExpression {
  const self: IFlatExpression = {
    context,
    collect: (): OptionalChainHelper => {
      let initialLeft;

      if (self.conditionSteps) {
        initialLeft = createPartialExpression({
          flatExpression: self,
          nodes: self.conditionSteps,
          shouldExtractThis: !!self.alternate[0].callArguments,
        });
      }

      self.id = context.genId();
      const targets = self.alternate;
      let expression = createOptionalChaningExpression(self.id);

      if (self.conditionSteps) {
        expression.setLeft(initialLeft.expression);
        if (initialLeft.thisVariable) {
          self.thisVariable = initialLeft.thisVariable;
        }
      }

      // inserting the current identifier

      targets.unshift({ expression: { name: self.id, type: 'Identifier' } });
      const dataRight = createPartialExpression({
        flatExpression: self,
        nodes: targets,
        shouldExtractThis: self.shouldExtractThis,
      });
      expression.setRight(dataRight.expression);

      // should notify the following call expression to use "this" variable
      if (self.nextFlatExpression && dataRight.thisVariable)
        self.nextFlatExpression.thisVariable = dataRight.thisVariable;

      return expression;
    },

    generate: (): OptionalChainHelper => {
      let expression = self.collect();
      let next = self.nextFlatExpression;
      while (next) {
        const data = next.collect();
        data.setLeft(expression.statement);
        expression = data;
        next = next.nextFlatExpression;
      }
      return expression;
    },
  };
  return self;
}

function processFlatCollection(context: OptionalChainContext) {
  const flatCollection = context.flatCollection;
  let current: IFlatExpression;

  let index = flatCollection.length - 1;
  while (index >= 0) {
    const item = flatCollection[index];
    if (index > 0) {
      let flatExpression: IFlatExpression = createFlatExpression(context);
      flatExpression.alternate = item.items;
      if (current) {
        flatExpression.shouldExtractThis = !!current.alternate[0].callArguments;
        flatExpression.nextFlatExpression = current;
      }
      current = flatExpression;
    } else {
      current.conditionSteps = item.items;
    }
    index--;
  }

  const { statement } = current.generate();
  context.finalStatement = statement;
}

function flatten(context: OptionalChainContext) {
  const { steps } = context;
  const amount = steps.length;

  const collection: IChainingCollection = [];

  let index = amount - 1;
  while (index >= 0) {
    const step = steps[index];
    const prev = collection[collection.length - 1];
    if (!step.optional && prev) prev.items.push(step);
    else collection.push({ items: [step] });
    index--;
  }
  context.flatCollection = collection;
  processFlatCollection(context);
}
/**
 * Drill every single property on the OptionalChain
 * Split it into steps and prepare for flattening
 * @param node
 * @param context
 */
export function chainDrill(node: ASTNode, context: OptionalChainContext) {
  const optional = node.optional === true;

  if (NON_DRILLABLE_NODES[node.type]) {
    return context.steps.push({ computed: node.computed, expression: node });
  }

  if (node.type === ASTType.OptionalMemberExpression) {
    if (node.property) context.steps.push({ computed: node.computed, expression: node.property, optional });
    if (node.object) {
      if (node.object.type === ASTType.CallExpression) {
        context.steps.push({ expression: node.object, optional });
      } else chainDrill(node.object, context);
    }
    return;
  }

  // Regardless of the "optional" status calls should be separated since the call to those
  if (node.type === ASTType.OptionalCallExpression) {
    if (node.callee) {
      context.steps.push({
        callArguments: node.arguments,
        optional,
      });
      chainDrill(node.callee, context);
    }
  }
}

export function OptionalChaningTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const node = visit.node;
          if (!VALID_NODES[node.type]) return;
          const context = createOptionalContext(visit);

          chainDrill(node, context);
          flatten(context);

          return { prependToBody: [context.declaration], replaceWith: context.finalStatement };
        },
      };
    },
  };
}
