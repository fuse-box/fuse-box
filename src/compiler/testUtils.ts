import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ImportType } from './interfaces/ImportType';
import { ITransformer } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { parseTypeScript } from './parser';
import { createGlobalContext } from './program/GlobalContext';
import { transpileModule } from './program/transpileModule';
import { GlobalContextTransformer } from './transformers/GlobalContextTransformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });

  // const ast = buntis.parseTSModule(props.code, {
  //   directives: true,
  //   jsx: props.jsx,
  //   next: true,
  //   loc: true,
  // });
  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors({ ...props.props, onRequireCallExpression }));
    }
  }
  transpileModule({
    ast: ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
