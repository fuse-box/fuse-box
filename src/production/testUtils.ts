import * as appRoot from 'app-root-path';
import { unlinkSync, writeFileSync, unlink } from 'fs';
import * as path from 'path';
import { ASTNode } from '../compiler/interfaces/AST';
import { ITransformer } from '../compiler/interfaces/ITransformer';
import { createGlobalContext } from '../compiler/program/GlobalContext';
import { transpileModule } from '../compiler/program/transpileModule';
import { GlobalContextTransformer } from '../compiler/transformers/GlobalContextTransformer';
import { IPublicConfig } from '../config/IPublicConfig';
import { createContext } from '../core/Context';
import { fusebox } from '../core/fusebox';
import { createModule } from '../core/Module';
import { createPackage } from '../core/Package';
import { ensureDir, fastHash } from '../utils/utils';
import { ModuleTree } from './module/ModuleTree';
import { IProductionContext, ProductionContext } from './ProductionContext';
import * as fsExtra from 'fs-extra';
export function testProductionWarmup(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
  moduleProps?: any;
}) {
  const ctx = createContext({
    cache: false,
    devServer: false,
    target: 'browser',
  });

  const pkg = createPackage({ ctx: ctx, meta: {} as any });

  const module = Object.assign(
    createModule(
      {
        absPath: __filename,
        ctx: ctx,
        extension: '.ts',
        fuseBoxPath: 'file.ts',
      },
      pkg,
    ),
    props.moduleProps || {},
  );

  pkg.modules.push(module);

  module.contents = props.code;
  module.parse();

  const productionContext = ProductionContext(ctx, [pkg]);

  const tree = (module.moduleTree = ModuleTree(productionContext, module));

  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.productionWarmupPhase) {
      tranformers.push(t.productionWarmupPhase({ ctx: ctx, module: module, productionContext: productionContext }));
    }
  }
  transpileModule({
    ast: module.ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });

  return { productionContext, module, tree };
}

function createFileSet(directory, files: Record<string, string>) {
  const newSet = {};
  for (const key in files) {
    newSet[path.join(directory, key)] = files[key];
  }

  return newSet;
}
export class ProdPhasesTestEnv {
  private rootDir: string;
  private homeDir: string;
  constructor(public config: IPublicConfig) {
    config.cache = false;
    config.target = 'browser';
    config.watch = false;
    config.devServer = false;
  }
  initFileStructure(files: Record<string, string>) {
    const folder = fastHash(Math.random() + '-' + new Date().getTime());
    this.rootDir = path.join(appRoot.path, '.tmp', folder);
    this.homeDir = path.join(this.rootDir, 'src');
    this.config.homeDir = this.homeDir;
    ensureDir(this.homeDir);
    for (const key in files) {
      const filePath = path.join(this.homeDir, key);
      writeFileSync(filePath, files[key]);
    }
  }

  public async run(phases): Promise<IProductionContext> {
    const fuse = fusebox(this.config);
    const context = await fuse.runProductionContext(phases);
    return context;
  }

  public destroy() {
    fsExtra.removeSync(this.rootDir);
  }
}

export function prodPhasesEnv(config: IPublicConfig, files: Record<string, string>): ProdPhasesTestEnv {
  const env = new ProdPhasesTestEnv(config);
  env.initFileStructure(files);

  return env;
}
