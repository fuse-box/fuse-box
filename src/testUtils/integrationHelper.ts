import { EnvironmentType } from '../config/EnvironmentType';
import { IPublicConfig } from '../config/IConfig';
import { ITestBrowserResponse } from './browserEnv/testBrowserEnv';
import { createTestWorkspace, testBrowser, testServer, ITestWorkspace } from './integrationTest';
import { ITestServerResponse } from './serverEnv/testServerEnv';

export const runBrowserTest = async (props: {
  config?: IPublicConfig;
  env: EnvironmentType;
  files: Record<string, string>;
  modules?: Record<string, Record<string, string>>;
}): Promise<{ response: ITestBrowserResponse; workspace: ITestWorkspace }> => {
  const workspace = createTestWorkspace({ files: props.files, modules: props.modules });

  const b = await testBrowser({ config: props.config || {}, type: props.env, workspace: workspace });
  return { response: b, workspace };
};

export const runServerTest = async (props: {
  config?: IPublicConfig;
  env: EnvironmentType;
  files: Record<string, string>;
  modules?: Record<string, Record<string, string>>;
}): Promise<{ response: ITestServerResponse; workspace: ITestWorkspace }> => {
  const workspace = createTestWorkspace({ files: props.files, modules: props.modules });

  const b = await testServer({ config: props.config || {}, type: props.env, workspace: workspace });
  return { response: b, workspace };
};
