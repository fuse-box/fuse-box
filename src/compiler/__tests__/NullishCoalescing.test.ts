import { initCommonTransform } from '../testUtils';
import { NullishCoalescingTransformer } from '../transformers/nullishCoalescing/NullishCoalescingTransformer';
import { OptionalChaningTransformer } from '../transformers/optionalChaining/OptionalChainingTransformer';

const testTranspile = (props: { code: string; jsx?: boolean }) => {
  return initCommonTransform({
    code: props.code,
    jsx: true,
    transformers: [NullishCoalescingTransformer(), OptionalChaningTransformer()],
  });
};

describe('NullishCoalescing', () => {
  it('case 1', () => {
    const result = testTranspile({
      code: `
         a ?? b
        `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('case 1 with obj', () => {
    const result = testTranspile({
      code: `
         a.c ?? d
        `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('case 1 with option call', () => {
    const result = testTranspile({
      code: `
         a.c() ?? d
        `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('long 1', () => {
    const result = testTranspile({
      code: `
         foo.bar ?? oi ?? soi ?? doi.o
        `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('long with optional chaning', () => {
    const result = testTranspile({
      code: `
         foo?.bar ?? oi ?? soi ?? doi?.o
        `,
    });
    expect(result.code).toMatchSnapshot();
  });
});
