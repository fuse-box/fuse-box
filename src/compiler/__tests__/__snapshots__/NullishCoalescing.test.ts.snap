// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NullishCoalescing case 1 1`] = `
"a !== null && a !== void 0 ? a : b;
"
`;

exports[`NullishCoalescing case 1 with obj 1`] = `
"var _1_;
(_1_ = a.c) !== null && _1_ !== void 0 ? _1_ : d;
"
`;

exports[`NullishCoalescing case 1 with option call 1`] = `
"var _1_;
(_1_ = a.c()) !== null && _1_ !== void 0 ? _1_ : d;
"
`;

exports[`NullishCoalescing long 1 1`] = `
"var _1_, _2_, _3_;
(_3_ = (_2_ = (_1_ = foo.bar) !== null && _1_ !== void 0 ? _1_ : oi) !== null && _2_ !== void 0 ? _2_ : soi) !== null && _3_ !== void 0 ? _3_ : doi.o;
"
`;

exports[`NullishCoalescing long with optional chaning 1`] = `
"var _1_, _2_, _3_;
var _4_;
var _5_;
(_3_ = (_2_ = (_1_ = (_5_ = foo) === null || _5_ === void 0 ? void 0 : _5_.bar) !== null && _1_ !== void 0 ? _1_ : oi) !== null && _2_ !== void 0 ? _2_ : soi) !== null && _3_ !== void 0 ? _3_ : (_4_ = doi) === null || _4_ === void 0 ? void 0 : _4_.o;
"
`;
